### ISSUE TITLE
Migrate exprfilter to modern JIT library

### ISSUE THREAD

**Issue by sekrit-twc** [vapoursynth/vapoursynth#539]
Problems with existing jitasm code:
- Abandonware. Last updated 2016(?)
- Doesn't support latest instruction sets (e.g. AVX-512)
- Doesn't support RIP-relative encoding: constants have to be stored as immediates, which require a round-trip through a GPR
- Can't call subroutines: leads to code-bloat in expressions with multiple exp/pow

Possible solutions:
- Migrate to asmjit, which is up-to-date: https://github.com/asmjit/asmjit
- Migrate to LLVM

**Comment by myrsloik** [comment]
Do it! LLVM seem like more fun...

**Comment by pinterf** [comment]
I considered using asmjit when found those jitasm bugs and seeing that even avx2 support needed some finishing steps. I'm sure it has shorter learning curve, but it's x86 only though unlike LLVM?
I had further enhancement ideas such as working on integer data or doing math on 64 bit doubles.

**Comment by sekrit-twc** [comment]
Hi pinterf,

IMO, asmjit would be the easier move, but LLVM has the benefit of potentially enabling moving beyond the MaskTools RPN language. With a LLVM/Clang backend, we could enable the user to execute arbitrary procedures as an inline C program.

**Comment by sekrit-twc** [comment]
Another idea that came up: convert to OpenCL and use a CPU-only runtime.

https://software.intel.com/en-us/iocl-rt-ref-tech-intel-immediate-command-execution

It looks like the Intel OpenCL driver for CPUs can execute OpenCL using "host" memory and without any async stuff, so it would map directly to a VSynth filter. Closed and proprietary though.

**Comment by Endilll** [comment]
Hi.
Just letting you know that I got the following code working as one would expect [here](https://github.com/Endilll/exprcpp):
```
user_func = '''
int func(int x, int y) {
    return x/4 + y/4;
}
'''
clip00 = core.endill.expr_cpp(clip00, clip00, user_func)
```
LLVM is really fun, except when it's not.

**Comment by myrsloik** [comment]
Fun indeed!
Your proof of concept always only takes 2 inputs, right?
I think the most interesting improvement to make would be to also move the iteration over all the pixels into the LLVM generated code and attach alignment/restrict hints to the inputs. If you're lucky you could get decent auto vectorization on the other side.

**Comment by Endilll** [comment]
> Your proof of concept always only takes 2 inputs, right?

Yes. I just haven't put much effort into supporting code so far. 

> to also move the iteration over all the pixels into the LLVM generated code

Yes, it's been on my mind, too.

> attach alignment/restrict hints to the inputs

Thanks for the suggestion.

Quick test suggests that current implementation is (just) 2 times slower than `Expr` on example from readme, which is better that I anticipated.

**Comment by sekrit-twc** [comment]
Is it possible to implement disassembly of the JIT code? For debugging.

**Comment by Endilll** [comment]
Quick answer is that it should be possible

**Comment by Endilll** [comment]
I have my ideas on how to improve current `Expr` interface, but I'd love to hear from you, too, if you have any. And about pitfalls I might fall into that are relevant even after replacing MaskTools RPN with C++, too.

**Comment by WolframRhodium** [comment]
I believe an approach based on operator-overloading is more readable ([example](https://github.com/WolframRhodium/muvsfunc/wiki/muvs-tutorial#Unified-infix-expression-syntax)). Generalizing JIT to broader cases, e.g. spatial processing seems even more interesting (maybe numba or [Halide](https://github.com/halide/Halide) would be helpful).

**Comment by sekrit-twc** [comment]
IMO, the JIT should aim first and foremost to optimize the performance of existing RPN expressions, which are widely used today. Unfortunately, C compilers, including Clang/LLVM are terrible at vectorization, even when all the requisite pre-conditions are met. IMO, it would be better to generate LLVM bytecode (perhaps directly vectorized), than to allow users to input C programs. 

As a motivating example, let us consider the expression "", which is far simpler than the most challenging expressions of havsfunc. Directly translating the expression to infix notation compiles to this garbage:

```
void f(const uint8_t *p1, const uint8_t *p2, uint8_t *p3, ptrdiff_t w, ptrdiff_t h, ptrdiff_t stride1, ptrdiff_t stride2, ptrdiff_t stride3)
{
    using std::min;
    using std::max;

    for (ptrdiff_t i = 0; i < h; ++i) {
        for (ptrdiff_t j = 0; j < w; ++j) {
            const uint8_t *__restrict pp1 = (const uint8_t *)__builtin_assume_aligned(p1, 32);
            const uint8_t *__restrict pp2 = (const uint8_t *)__builtin_assume_aligned(p2, 32);
            uint8_t *__restrict pp3 = (uint8_t *)__builtin_assume_aligned(p3, 32);

            float x = pp1[j];
            float y = pp2[j];
            float result = fabsf(x - y) > 4 ? x * y : x * x;
            pp3[j] = (uint8_t)min(max(lrintf(result), 0L), 255L);
        }

        p1 += stride1;
        p2 += stride2;
        p3 += stride3;
    }
}

.LBB0_7:                                #   in Loop: Header=BB0_3 Depth=1
        add     rdi, r9
        add     rsi, r11
        add     rdx, r10
        inc     r14
        cmp     r14, r8
        je      .LBB0_8
.LBB0_3:                                # =>This Loop Header: Depth=1
        xor     ebx, ebx
        cmp     rcx, 1
        je      .LBB0_5
.LBB0_4:                                #   Parent Loop BB0_3 Depth=1
        movzx   eax, byte ptr [rdi + rbx]
        vcvtsi2ss       xmm3, xmm6, eax
        movzx   eax, byte ptr [rsi + rbx]
        vcvtsi2ss       xmm4, xmm6, eax
        vsubss  xmm5, xmm3, xmm4
        vandps  xmm5, xmm5, xmm2
        vcmpltss        xmm5, xmm1, xmm5
        vblendvps       xmm4, xmm3, xmm4, xmm5
        vmulss  xmm3, xmm4, xmm3
        vcvtss2si       rbp, xmm3
        mov     rax, rbp
        sar     rax, 63
        andn    rax, rax, rbp
        cmp     rax, 255
        cmovge  eax, r12d
        mov     byte ptr [rdx + rbx], al
        movzx   eax, byte ptr [rdi + rbx + 1]
        vcvtsi2ss       xmm3, xmm6, eax
        movzx   eax, byte ptr [rsi + rbx + 1]
        vcvtsi2ss       xmm4, xmm6, eax
        vsubss  xmm5, xmm3, xmm4
        vandps  xmm5, xmm5, xmm2
        vcmpltss        xmm5, xmm1, xmm5
        vblendvps       xmm4, xmm3, xmm4, xmm5
        vmulss  xmm3, xmm4, xmm3
        vcvtss2si       rbp, xmm3
        mov     rax, rbp
        sar     rax, 63
        andn    rax, rax, rbp
        cmp     rax, 255
        cmovge  eax, r12d
        mov     byte ptr [rdx + rbx + 1], al
        add     rbx, 2
        cmp     r15, rbx
        jne     .LBB0_4
.LBB0_5:                                #   in Loop: Header=BB0_3 Depth=1
        test    cl, 1
        je      .LBB0_7
        movzx   eax, byte ptr [rdi + rbx]
        vcvtsi2ss       xmm3, xmm6, eax
        movzx   eax, byte ptr [rsi + rbx]
        vcvtsi2ss       xmm4, xmm6, eax
        vsubss  xmm5, xmm3, xmm4
        vandps  xmm5, xmm5, xmm0
        vcmpltss        xmm5, xmm1, xmm5
        vblendvps       xmm4, xmm3, xmm4, xmm5
        vmulss  xmm3, xmm4, xmm3
        vcvtss2si       rbp, xmm3
        mov     rax, rbp
        sar     rax, 63
        andn    rax, rax, rbp
        cmp     rax, 255
        cmovge  eax, r12d
        mov     byte ptr [rdx + rbx], al
        jmp     .LBB0_7

```

Translating the expression directly to AVX2 intrinsics, without even bother to apply simple arithmetic and common-value optimizations produces essentially optimal code, even better than what R48's optimizer achieves.

```
void g(const uint8_t *p1, const uint8_t *p2, uint8_t *p3, ptrdiff_t w, ptrdiff_t h, ptrdiff_t stride1, ptrdiff_t stride2, ptrdiff_t stride3)
{
    using std::min;
    using std::max;

#define SPLAT(x) { (x), (x), (x), (x), (x), (x), (x), (x) }
    alignas(32) static const uint32_t negmask[8] = SPLAT(0x80000000);
    alignas(32) static const float one[8] = SPLAT(1.0f);
    alignas(32) static const float zero[8] = SPLAT(0.0f);
    alignas(32) static const float float_255[8] = SPLAT(255.0f);
#undef SPLAT

    for (ptrdiff_t i = 0; i < h; ++i) {
        for (ptrdiff_t j = 0; j < w; ++j) {
            const uint8_t *__restrict pp1 = (const uint8_t *)__builtin_assume_aligned(p1, 32);
            const uint8_t *__restrict pp2 = (const uint8_t *)__builtin_assume_aligned(p2, 32);
            uint8_t *__restrict pp3 = (uint8_t *)__builtin_assume_aligned(p3, 32);

            __m256 v1 = _mm256_cvtepi32_ps(_mm256_cvtepu8_epi32(_mm_loadl_epi64((const __m128i*)(pp1 + j + 8))));
            __m256 v2 = _mm256_cvtepi32_ps(_mm256_cvtepu8_epi32(_mm_loadl_epi64((const __m128i*)(pp2 + j + 8))));
            __m256 v3 = _mm256_sub_ps(v1, v2);
            __m256 v4 = _mm256_and_ps(v3, _mm256_load_ps((const float *)negmask));
            __m256 v5 = _mm256_set1_ps(4.0f);
            __m256 v6 = _mm256_and_ps(_mm256_cmp_ps(v4, v5, _CMP_NLE_US), _mm256_load_ps(one));
            __m256 v7 = _mm256_cvtepi32_ps(_mm256_cvtepu8_epi32(_mm_loadl_epi64((const __m128i*)pp1 + j + 8)));
            __m256 v8 = _mm256_cvtepi32_ps(_mm256_cvtepu8_epi32(_mm_loadl_epi64((const __m128i*)pp2 + j + 8)));
            __m256 v9 = _mm256_mul_ps(v7, v8);
            __m256 v10 = _mm256_cvtepi32_ps(_mm256_cvtepu8_epi32(_mm_loadl_epi64((const __m128i*)pp1 + j + 8)));
            __m256 v11 = _mm256_cvtepi32_ps(_mm256_cvtepu8_epi32(_mm_loadl_epi64((const __m128i*)pp1 + j + 8)));
            __m256 v12 = _mm256_mul_ps(v10, v11);
            __m256 v13 = _mm256_blendv_ps(v12, v9, _mm256_cmp_ps(v6, _mm256_load_ps(zero), _CMP_NLE_US));

            __m256 resultf = _mm256_min_ps(v13, _mm256_load_ps(float_255));
            __m256i result = _mm256_cvtps_epi32(v13);
            result = _mm256_packus_epi32(result, result);
            result = _mm256_permute4x64_epi64(result, 0x08);
            _mm_store_si128((__m128i *)(pp3 + j), _mm256_castsi256_si128(result));
        }

        p1 += stride1;
        p2 += stride2;
        p3 += stride3;
    }
}

.LBB1_3:                                # =>This Loop Header: Depth=1
        mov     ebx, 128
        xor     eax, eax
.LBB1_4:                                #   Parent Loop BB1_3 Depth=1
        vpmovzxbd       ymm2, qword ptr [rdi + rax + 8] # ymm2 = mem[0],zero,zero,zero,mem[1],zero,zero,zero,mem[2],zero,zero,zero,mem[3],zero,zero,zero,mem[4],zero,zero,zero,mem[5],zero,zero,zero,mem[6],zero,zero,zero,mem[7],zero,zero,zero
        vcvtdq2ps       ymm2, ymm2
        vpmovzxbd       ymm3, qword ptr [rsi + rax + 8] # ymm3 = mem[0],zero,zero,zero,mem[1],zero,zero,zero,mem[2],zero,zero,zero,mem[3],zero,zero,zero,mem[4],zero,zero,zero,mem[5],zero,zero,zero,mem[6],zero,zero,zero,mem[7],zero,zero,zero
        vcvtdq2ps       ymm3, ymm3
        vsubps  ymm2, ymm2, ymm3
        vandps  ymm2, ymm2, ymm0
        vcmpltps        ymm2, ymm1, ymm2
        vpmovzxbd       ymm3, qword ptr [rdi + rbx] # ymm3 = mem[0],zero,zero,zero,mem[1],zero,zero,zero,mem[2],zero,zero,zero,mem[3],zero,zero,zero,mem[4],zero,zero,zero,mem[5],zero,zero,zero,mem[6],zero,zero,zero,mem[7],zero,zero,zero
        vcvtdq2ps       ymm3, ymm3
        vpmovzxbd       ymm4, qword ptr [rsi + rbx] # ymm4 = mem[0],zero,zero,zero,mem[1],zero,zero,zero,mem[2],zero,zero,zero,mem[3],zero,zero,zero,mem[4],zero,zero,zero,mem[5],zero,zero,zero,mem[6],zero,zero,zero,mem[7],zero,zero,zero
        vcvtdq2ps       ymm4, ymm4
        vblendvps       ymm2, ymm3, ymm4, ymm2
        vmulps  ymm2, ymm2, ymm3
        vcvtps2dq       ymm2, ymm2
        vpackusdw       ymm2, ymm2, ymm0
        vextracti128    xmm3, ymm2, 1
        vpunpcklqdq     xmm2, xmm2, xmm3 # xmm2 = xmm2[0],xmm3[0]
        vmovdqa xmmword ptr [rdx + rax], xmm2
        inc     rax
        add     rbx, 16
        cmp     rcx, rax
        jne     .LBB1_4
        add     rdi, r9
        add     rsi, r11
        add     rdx, r10
        inc     r14
        cmp     r14, r8
        jne     .LBB1_3
```

See on Godbolt: https://godbolt.org/z/jBYvjc

tl;dr: Constricting user inputs to enable the compiler to produce fast code is more important than having supporting the most expressive code. A first step could be to take the existing `exprfilter.cpp` and change the backend from jitasm to LLVM vector IR. Some investigation can be done to see if that IR can be vector length agnostic, or if separate backends for SSE, AVX, and AVX-512 are required.

**Comment by Endilll** [comment]
Thanks to both of you for the feedback!

@WolframRhodium 
You have very interesting stuff happening there that I haven't even considered. Currently I don't see a way to accommodate the approach used there, but thank you for letting me know. I also should say that having elimination of common sub-expressions is impressive.

@sekrit-twc 
1. Let me clarify my motivation for starting ExprCpp project: (a) I'm interested in Clang and LLVM; (b) RPN is anything but expressive; (c) current implementation of Expr doesn't perform elimination of sub-expressions (as far as I can tell).
2. Expr needs to serve non-expert users as well as ones who can reason about code with stack manipulation functions with confidence. And it's to be expected that those users are not bottlenecked by Expr, so they can trade some performance for expressiveness. I'm also sure that there're experts that would like to have something more expressive for prototyping.
3. After taking a glance at LLVM IR reference and IR bitcode emitted by Clang in your example, I have a feeling that it's frontend that generates suboptimal output. It goes without saying that it needs further investigation.
4. I think RPN is a dead end as a language. It may be as fast as carefully written intrinsic code, but does it really matter for [this user](https://github.com/vapoursynth/vapoursynth/issues/541), as well as others who need to go beyond simple arithmetic and few more functions Expr has to offer? Don't get me wrong: performance matters. But libraries matter as well.
5. RPN can't be simply abandoned because of all the code being written so far. It would require writing a frontend, and it's something I consider because of (a). Since it'll be under our full control, we'll be able to circumvent (3), if I'm not mistaken there. But as for now, it's in the lower half of my priority list.

As a bottom line, I don't expect ExprCpp being able to completely replace Expr in foreseeable future, so I don't want to discourage anyone who'd like to invest into it, raising performance baseline for ExprCpp among the other improvements.

**Comment by sekrit-twc** [comment]
@Endilll 

1-2. std.Expr performs CSE, strength reduction, and a number of common optimizations starting with R48. Likewise, it is no longer needed (or beneficial) for users to use the `dup` and `swap` operators.

3-5. I agree that RPN is a dead-end language, but having (a) good performance for existing scripts, and (b) making it simple for users to write new scripts that still perform well are essential. That is why I feel a general C/C++ language frontend is not ideal, because even in the code I presented, there is some language-related issue preventing vectorization. A compromise that should be investigated is whether a different frontend to LLVM can produce better IR, such as OpenCL.

**Comment by Endilll** [comment]
Adding to C++ disadvantages, I have to admit that it can't beat simple limiter like `x if x > 20 else 0` because of function signature boilerplate (kind of). As for presumably suboptimal IR produced by Clang, I wouldn't rule out the possibility of fixing it upstream.

OpenCL is something I'd like to leave to somebody else, because currently I'm not interested much in it.

**Comment by Endilll** [comment]
@sekrit-twc 
https://godbolt.org/z/UJdq9e
I took your example and rewrote it for ExprCpp (source #1). Without mistakes, hopefully.
I can't really understand intrinsics assembly, but it looks promising on 10, and seems to improve on trunk.

But I can't say the same about output of JIT (IR source and assembly source). It's exact the same source, and exact the same flags were supplied to Clang driver (with the addition of `-fsyntax-only` to make things work). I'm also surprised that it doesn't pick target microarchitecture by default: it falls back to SSE without `-march=skylake`. It makes me think that I don't get everything out of Clang frontend. I'll work on this.

As it is, ExprCpp is about 30% slower than Expr on this example on my machine (i7-8656U).

**Comment by sekrit-twc** [comment]
I took a look at your example. It seems the issue is with int<-->float conversions. When I change the data type of `func` to take and return floats, then Clang/LLVM once again fail to vectorize. It is worth investigating if the issue lies in Clang or in LLVM.

**Comment by Endilll** [comment]
@sekrit-twc I created a new release, which supports dumping of C++ source, LLVM bitcode and native binary, as well as custom compiler flags, if you're interested in checking out it's capabilities.

Early testing suggests that it can be as fast as Expr. More interestingly, unvectorized code seems to work just 15% slower on my machine than Expr.

**Comment by myrsloik** [comment]
Not that surprising. In a shorter expression you spend maybe 1/3 of the instructions convertion to/from the source/destination format which can be skipped and save lots of time. And another 1/3 ends up being jumps and reads that can be avoided when compiling. You get the point. If a compiler generated code as inefficient you'd cry...

**Comment by Endilll** [comment]
Speaking of GPU support, it seems that ExprCpp could be further developed to use DPC++ compiler, which is based on LLVM. Glue code that goes inside JIT will obviously require a rewrite in SYCL, but the best part of this plan is that I don't see any fundamental obstacles for supporting code users are writing now (if there're any).

The problem is that DPC++ has a much more complicated pipeline, and after spending few hours inside it's source code, I wasn't able to found basic blocks of that pipeline to reuse.

**Comment by AkarinVS** [comment]
I've created a PoC LLVM based `std.Expr` implementation.

Currently it's only able to do 4-way SIMD, and for the expression mentioned above `x y - abs 4 > x y * x x * ?`,
it generates the following code:
```
   0x7ffff7ff5000:      mov    (%rdi),%r8
   0x7ffff7ff5003:      mov    0x8(%rdi),%rcx
   0x7ffff7ff5007:      mov    0x10(%rdi),%rsi
   0x7ffff7ff500b:      xor    %edi,%edi
   0x7ffff7ff500d:      movabs $0x7ffff7ff4000,%rax
   0x7ffff7ff5017:      vmovaps (%rax),%xmm0
   0x7ffff7ff501b:      movabs $0x7ffff7ff4010,%rax
   0x7ffff7ff5025:      vmovaps (%rax),%xmm1
   0x7ffff7ff5029:      xor    %eax,%eax
   0x7ffff7ff502b:      cmp    %edx,%eax
   0x7ffff7ff502d:      jge    0x7ffff7ff5061
   0x7ffff7ff502f:      nop
   0x7ffff7ff5030:      vmovups (%rcx,%rdi,1),%xmm2
   0x7ffff7ff5035:      vmovups (%rsi,%rdi,1),%xmm3
   0x7ffff7ff503a:      vsubps %xmm3,%xmm2,%xmm4
   0x7ffff7ff503e:      vandps %xmm0,%xmm4,%xmm4
   0x7ffff7ff5042:      vcmpltps %xmm4,%xmm1,%xmm4
   0x7ffff7ff5047:      vblendvps %xmm4,%xmm3,%xmm2,%xmm3
   0x7ffff7ff504d:      vmulps %xmm3,%xmm2,%xmm2
   0x7ffff7ff5051:      vmovups %xmm2,(%r8,%rdi,1)
   0x7ffff7ff5057:      inc    %eax
   0x7ffff7ff5059:      add    $0x10,%rdi
   0x7ffff7ff505d:      cmp    %edx,%eax
   0x7ffff7ff505f:      jl     0x7ffff7ff5030
   0x7ffff7ff5061:      retq
```

Except for being 4-way SIMD (AVX1), the code is very optimal. In fact, benchmark against the current 8-way AVX2 based `std.Expr` showed that it is ~10% faster on 1080p BlankClips. Of course, JIT compilation itself will be slower, but that's not a big deal for our intended use cases and not to mention LLVM is also able to address some of the shortcomings of the current compiler (e.g. exponential space blowup for extensive `dup` usage, for a dramatic example, see [muvsfunc_misc.fast_mandelbrot](https://github.com/WolframRhodium/muvsfunc/blob/c72143dc0151a9a82f96f2f8356d08d5ff7d84b8/Collections/muvsfunc_misc.py#L570), LLVM is able to do `iterations=10000` in 1 minute, but `std.Expr` struggles even with `iterations=5`.)

There is some issues worth discussion before creating a PR:
1. LLVM is licensed Apache 2 and it's incompatible with our LGPL v2.1. Even though LLVM has an exception for GPLv2, I'm not sure if it covers LGPL v2.1 as well. I'd propose we migrate to LGPL v3 to clear this hurdle once and for all.

2. LLVM does not have _stable_ API and it's a huge library (e.g. my current implementation is ~20MB large when built with VS2017 for x64 windows.) If we're going to adopt this approach, how should we deal with these issues?
   1. Should we vendor LLVM? (Note that the upstream LLVM does not have a single git repository for llvm itself, and the https://github.com/llvm/llvm-project repository is around 1GB, so I doubt we want to use submodule, at least not against the official llvm-project repository.
   2. Or we can just ask the user to install whatever LLVM version we require. (Judging by other projects' experiences, this is likely not very practical.)

3. If we choose to somehow build LLVM along with the rest of VS, then there are some non-negligible engineering required to be able to build LLVM using autotools, as its native build tool is cmake.

4. Do we still need an interpreter? LLVM has a very wide range of supported architectures, so I'm not sure it's worth the trouble to provide an interpreter.

Thanks. Please let me know your thoughts. Of course, I'm also happy to maintain my implementation as a separate plugin and I kinda think adding 20MB dependencies for `std.Expr` is probably too overkill.

**Comment by myrsloik** [comment]
Definitely needs to be a separate plugin. LLVM is an insane dependency to have. If you market its efficiency scriptwriters will conditionally use it when available (see eedi3 and eedi3m).

**Comment by sekrit-twc** [comment]
While I would love to see lexpr in VS, I agree that depending on LLVM isn't practical. Can LLVM be easily extended to YMM and ZMM operation?

**Comment by AkarinVS** [comment]
On Mon, Jun 14, 2021 at 5:49 PM sekrit-twc ***@***.***> wrote:

> While I would love to see lexpr in VS, I agree that depending on LLVM
> isn't practical. Can LLVM be easily extended to YMM and ZMM operation?
>
It's currently emitting 4-way vector types, but there is no limitation on
using a wider vector type.
(All ops are polymorphic, so it should be as simple as changing a few
constants and types.)

In fact, I'm looking at scalable vector types so that it can make good use
of ARM SVE and RISV-V Vector extensions.

If you want to give it a try, I made available a binary release for
windows/x64 here:
https://github.com/AkarinVS/vapoursynth-plugin/releases/tag/wip-v2

I will move it to a new plugin called lexpr. Thanks for the name
suggestions.

**Comment by myrsloik** [comment]
Keep us updated when you have a build ready for the world so I can add it to vsrepo.

**Comment by sekrit-twc** [comment]
Regarding LLVM, the previous contributor ran into problems with the vectorizer when dealing with int-float promotions. Does lexpr avoid this issue? The code snippit above appears to be for floating-point clips. Additionally, does lexpr successfully avoid branches being generated in expressions with deeper ternary trees?

**Comment by theChaosCoder** [comment]
> Keep us updated when you have a build ready for the world so I can add it to vsrepo.

It's already in the repo 😅
https://github.com/vapoursynth/vsrepo/blob/master/local/akarin.json

**Comment by AkarinVS** [comment]
Here are the 39 instructions generated for 10-bit int clips for the
benchmark expression: x y - abs 4 > x y * x x * ?
0x101f0a000: 4c 8b 07                       movq   (%rdi), %r8
0x101f0a003: 48 8b 4f 08                    movq   0x8(%rdi), %rcx
0x101f0a007: 48 8b 77 10                    movq   0x10(%rdi), %rsi
0x101f0a00b: 31 ff                          xorl   %edi, %edi
0x101f0a00d: 48 b8 20 b0 f0 01 01 00 00 00  movabsq $0x101f0b020, %rax ;
0x7fffffff
0x101f0a017: c4 e2 7d 18 00                 vbroadcastss (%rax), %ymm0
0x101f0a01c: 48 b8 24 b0 f0 01 01 00 00 00  movabsq $0x101f0b024, %rax ; 4.0
0x101f0a026: c4 e2 7d 18 08                 vbroadcastss (%rax), %ymm1
0x101f0a02b: c5 e8 57 d2                    vxorps %xmm2, %xmm2, %xmm2
0x101f0a02f: 48 b8 28 b0 f0 01 01 00 00 00  movabsq $0x101f0b028, %rax ;
1023.0
0x101f0a039: c4 e2 7d 18 18                 vbroadcastss (%rax), %ymm3
0x101f0a03e: 48 b8 00 b0 f0 01 01 00 00 00  movabsq $0x101f0b000, %rax ;
[0x00 0x01 0x04 0x05 0x08 0x09 0x0c 0x0d 0x08 0x09 0x0c 0x0d 0x0c 0x0d 0x0e
0x0f]
0x101f0a048: c5 fd 6f 20                    vmovdqa (%rax), %ymm4
0x101f0a04c: 31 c0                          xorl   %eax, %eax
0x101f0a04e: 39 d0                          cmpl   %edx, %eax
0x101f0a050: 7d 60                          jge    0x101f0a0b2
0x101f0a052: 66 2e 0f 1f 84 00 00 00 00 00  nopw   %cs:(%rax,%rax)
0x101f0a05c: 0f 1f 40 00                    nopl   (%rax)
0x101f0a060: c4 e2 7d 33 2c 39              vpmovzxwd (%rcx,%rdi), %ymm5
0x101f0a066: c5 fc 5b ed                    vcvtdq2ps %ymm5, %ymm5
0x101f0a06a: c4 e2 7d 33 34 3e              vpmovzxwd (%rsi,%rdi), %ymm6
0x101f0a070: c5 fc 5b f6                    vcvtdq2ps %ymm6, %ymm6
0x101f0a074: c5 d4 5c fe                    vsubps %ymm6, %ymm5, %ymm7
0x101f0a078: c5 c4 54 f8                    vandps %ymm0, %ymm7, %ymm7
0x101f0a07c: c5 f4 c2 ff 01                 vcmpltps %ymm7, %ymm1, %ymm7
0x101f0a081: c4 e3 55 4a f6 70              vblendvps %ymm7, %ymm6, %ymm5,
%ymm6
0x101f0a087: c5 cc 59 ed                    vmulps %ymm5, %ymm6, %ymm5
0x101f0a08b: c5 d4 5f ea                    vmaxps %ymm2, %ymm5, %ymm5
0x101f0a08f: c5 d4 5d eb                    vminps %ymm3, %ymm5, %ymm5
0x101f0a093: c5 fd 5b ed                    vcvtps2dq %ymm5, %ymm5
0x101f0a097: c4 e2 55 00 ec                 vpshufb %ymm4, %ymm5, %ymm5
0x101f0a09c: c4 e3 fd 00 ed e8              vpermq $0xe8, %ymm5, %ymm5
0x101f0a0a2: c4 c1 79 7f 2c 38              vmovdqa %xmm5, (%r8,%rdi)
0x101f0a0a8: ff c0                          incl   %eax
0x101f0a0aa: 48 83 c7 10                    addq   $0x10, %rdi
0x101f0a0ae: 39 d0                          cmpl   %edx, %eax
0x101f0a0b0: 7c ae                          jl     0x101f0a060
0x101f0a0b2: c5 f8 77                       vzeroupper
0x101f0a0b5: c3                             retq

The 35 instructions generated by our existing code is:
0x12efc8000: 55                             pushq  %rbp
0x12efc8001: 48 8b ec                       movq   %rsp, %rbp
0x12efc8004: c5 fd ef c0                    vpxor  %ymm0, %ymm0, %ymm0
0x12efc8008: 48 b8 20 20 35 03 01 00 00 00  movabsq $0x103352020, %rax
0x12efc8012: 48 8b 57 08                    movq   0x8(%rdi), %rdx
0x12efc8016: c4 e2 7d 33 0a                 vpmovzxwd (%rdx), %ymm1
0x12efc801b: c5 fc 5b c9                    vcvtdq2ps %ymm1, %ymm1
0x12efc801f: 48 8b 57 10                    movq   0x10(%rdi), %rdx
0x12efc8023: c4 e2 7d 33 12                 vpmovzxwd (%rdx), %ymm2
0x12efc8028: c5 fc 5b d2                    vcvtdq2ps %ymm2, %ymm2
0x12efc802c: c5 f4 5c da                    vsubps %ymm2, %ymm1, %ymm3
0x12efc8030: c5 e4 54 18                    vandps (%rax), %ymm3, %ymm3
0x12efc8034: ba 00 00 80 40                 movl   $0x40800000, %edx
0x12efc8039: c5 f9 6e e2                    vmovd  %edx, %xmm4
0x12efc803d: c4 e2 7d 18 ec                 vbroadcastss %xmm4, %ymm5
0x12efc8042: c5 e4 5c dd                    vsubps %ymm5, %ymm3, %ymm3
0x12efc8046: c5 f4 59 d2                    vmulps %ymm2, %ymm1, %ymm2
0x12efc804a: c5 f4 59 c9                    vmulps %ymm1, %ymm1, %ymm1
0x12efc804e: c5 e4 c2 e0 06                 vcmpnleps %ymm0, %ymm3, %ymm4
0x12efc8053: c4 e3 75 4a ca 40              vblendvps %ymm4, %ymm2, %ymm1,
%ymm1
0x12efc8059: c5 f4 5d 90 20 01 00 00        vminps 0x120(%rax), %ymm1, %ymm2
0x12efc8061: c5 fd 5b d2                    vcvtps2dq %ymm2, %ymm2
0x12efc8065: c4 e2 6d 2b d2                 vpackusdw %ymm2, %ymm2, %ymm2
0x12efc806a: c4 e3 fd 00 d2 08              vpermq $0x8, %ymm2, %ymm2
0x12efc8070: 48 8b 17                       movq   (%rdi), %rdx
0x12efc8073: c5 f8 29 12                    vmovaps %xmm2, (%rdx)
0x12efc8077: c5 fe 6f 0f                    vmovdqu (%rdi), %ymm1
0x12efc807b: c5 fe 6f 16                    vmovdqu (%rsi), %ymm2
0x12efc807f: c5 f5 d4 ca                    vpaddq %ymm2, %ymm1, %ymm1
0x12efc8083: c5 fe 7f 0f                    vmovdqu %ymm1, (%rdi)
0x12efc8087: 48 83 e9 01                    subq   $0x1, %rcx
0x12efc808b: 75 85                          jne    0x12efc8012
0x12efc808d: 5d                             popq   %rbp
0x12efc808e: c5 f8 77                       vzeroupper
0x12efc8091: c3                             retq

The differences are inconsequential (LLVM uses a different way to store
uint16_t and also adds two NOP instructions to pad loop body to be 16-byte
aligned.)

Regarding deeply nested ternary trees, the way I'm doing is that I do not
rely on the LLVM vectorize to vectorize the loop, but directly generate
expressions with explicit vector types, so ? will be translated roughly
into this:
(condIntVector & trueValue) | (~condIntVector & falseValue) just like the
current asmjit based backend, and LLVM is smart enough to use vblend if
supported.

Therefore, the code quality should be at least as good as the existing
backend, but the LLVM backend can handle significantly more complicated
cases.

**Comment by sekrit-twc** [comment]
So it directly generates vector LLVM IR? Sounds great! Especially if it can move sin/exp/pow into actual functions instead of copy-pasting them multiple times.

**Comment by myrsloik** [comment]
It's a separate plugin/project so I guess this issue has outlived its usefulness
